### ISSUE TITLE
Scripted tests takes too long to finish in private agent

### ISSUE THREAD

**Issue by abhagupta** [WPO-Foundation/webpagetest#1160]
We have noticed that a scripted test (which traverses between 6-7 steps) takes about avg. 35-40 mins in our private instance and eventually fails with no results produced.  I run the same same on public webpagetest instance, and it runs fine (avg time around 10 mins).  What could be reasons and how can we troubleshoot?  We have a Standard D4s_v3 (4 Cores, 16 GiB memory) machine in azure cloud.  

We have also seen that sometimes Test finishes up and agent is free to accept new tests, but test results keeps saying "Test Started X minutes ago"   Any ideas?

**Comment by zeman** [comment]
We see the same issue with Windows tests taking a long time. The public tests often run on Linux which don't seems to have the same issue when there are numerous steps. I'd suggest double checking your scripts and tuning them to run faster on Windows. Often the "execAndWait" command is used unnecessarily which can lead to the very long timeouts.

There's also an [open issue](https://github.com/WPO-Foundation/wptagent/issues/120) for providing a max time for a test so that at least it terminates when expected and doesn't tie up the agent unnecessarily.

**Comment by pmeenan** [comment]
Do you have a sample test to look at?  Is the test a single run, first-view only?  If not it is repeating the 6-7 steps multiple times.  If any of the steps in the script are timing out because they aren't working as expected each step could take 1-2 minutes (or if you are testing a 3G connection).  7 steps, 3 runs is 21 actual measurements which could easily get into the 35-40 minute range.

Mark linked to an issue that will help keep agents from being tied up but it won't actually help the test itself go any faster.  For that you'd need to look at the actual script on the browser being tested and make sure all of the steps are completing fast.

**Comment by pmeenan** [comment]
btw, the public instance may run it faster if there are several runs because it will shard the test across agents and run all of the runs in parallel (which explains 10 minutes vs 35-40 minutes).

**Comment by abhagupta** [comment]
Sample test : https://www.webpagetest.org/result/180712_9A_2b9234f6fcfe0790ec4c4198a99646d4/

Its first view only.

**Comment by abhagupta** [comment]
Please note that a lot of the steps in it are just doing actions on the same page.. That is why I mentioned 6-7 steps and not 11 steps (as shown in  test)

**Comment by pmeenan** [comment]
As long as there is a "andWait" the agent captures and processes video and waits for 2 seconds of no network activity.  That said, testing locally the whole sequence took around 2 minutes and that's around how long the test took.

One way to get more detail is to put a "combineSteps" at the beginning of the script.  That will run the entire sequence as a single step and you'll see any long points clearly.  You can also run with &debug=1 as a test option and the agents will capture a debug log for each step.

Do you see it taking 10 (or 30) minutes from the time the test started until it completed or from the time you submit the test until you get the result?  Just checking to see if there is some queuing time in what you are experiencing (or time for agents to start/stop).

**Comment by abhagupta** [comment]
I see 10 mins from test started until its completed.  Yes, you are right, the actual tests only takes 2-4 mins to complete. But the test sits on the last page for almost 15-20 mins avg ( Lately I have seen improvement).

I am making sure that there is no queueing up since I watch the test and it starts right away.
I will put your suggestions in place and try it out.. thanks

**Comment by pmeenan** [comment]
If it sits on the last page for that long. the last execAndWait command may not be working (though even then, no single step should ever take longer than 2 minutes.  If you collect a debug log it may help.

The only other thing that comes to mind is that it uploads the results all at once after the last step (possibly before closing the browser).  If the agent has a slow connection to the server it could be taking a long time to do the upload (unlikely but possible).

If you can connect to the agent while it is running you can see the activity in the console to see what it is doing.  The upload happens outside of the debug log so if nothing is visible in the debug log it could be that.

Another possibility is if the uploads are too big and the web server can't store the zip file in the temp directory (or it exceeds the overall POST size for php or the web server) then it is possible for the test to finish but for the server to not get the completion (you'd probably get an error on the test in that case though).

**Comment by abhagupta** [comment]
Thanks Patrick. Your suggestions helped a bit.   I had `logData 1`  which was keeping the test to sit at the last step. Removing that made the test complete on time but the status of the test is still "running " for almost 30 mins.   Not sure if this is where it is transferring the data and is probably failing.   if it is, how can I see network logs in order to verify?   How can I check the upload files from agent to server?

**Comment by abhagupta** [comment]
where do I see logs when we add  debug=1  .  `viewlog.php?test=180713_XG_15V`  does not give me anything more than usual.   My request is 

```
var request = require("request");

var options = { method: 'POST',
  url: '<myurl>',
  headers: 
   { 'postman-token': '4df63d70-b8ce-241c-1f98-4127fb27bef8',
     'cache-control': 'no-cache',
     'content-type': 'multipart/form-data; boundary=----WebKitFormBoundary7MA4YWxkTrZu0gW' },
  formData: 
   { script: <myscript>,
     location: 'CentralUS6:Chrome',
     debug: '1',
     fvonly: '1' } };

request(options, function (error, response, body) {
  if (error) throw new Error(error);

  console.log(body);
});
```

**Comment by abhagupta** [comment]
I am getting much better results now but I also see results like these:  
![image](https://user-images.githubusercontent.com/825965/42705306-bd28a7d4-8687-11e8-9e4f-eb8c11352e11.png)

(notice the missing first bytes and some other metrics.  I repeat the test, or run in different browser, I get these  values but miss out on others.. how can we get reliable tests results?

**Comment by pmeenan** [comment]
Check the screenshots and waterfalls to make sure it is doing what you expect it to do and that there is actually network activity being measured.

The "start render" missing for 2 of the steps is likely because nothing visually changed on the page.  User time missing is because those steps didn't log any performance.mark() events during the activity.  Missing document complete is likely because that step didn't trigger a page navigation.

**Comment by abhagupta** [comment]
Screenshots look good.  Agreed for start render and other metrics, but wondering why First Byte is missing.  Should I add some `sleep` to let system capture metrics at every step ?

Also, If I use `combineSteps`, then the values received are aggregate of all the steps happening in the tests, or are they measures of just the last step?  My question is more around calculation of  'customer journey'.  If `combineSteps` is actually an aggregate then thats what we may actually need and I do not have to worry about individual First Byte for every step.

**Comment by tkadlec** [comment]
We're in the process of cleaning up issues on this project in order to ensure we're able to stay on top of high priority bugs and feature requests. As a part of this process, we're automatically closing any issues that have had no activity within the last two years, including this one, since the codebase has changed dramatically in that time. If you feel this is still relevant, please [file a new issue](https://github.com/WPO-Foundation/webpagetest/issues/new/choose) using the relevant issue template so we can get it prioritized. Thanks!
